### PR TITLE
DateTimeWidget

### PR DESCRIPTION
**Pull request by zwim**
Deduplicate code by combining `datewidget` and `timewidget` into `datetimewidget.lua`.

Add `extra_buttons` to `datetimewidget.lua`

Replace all occurrences of `datewidget` and `timewidget` with `datetimewidget`

Reword misleading `Minutes form now` to `Time from now` in `readtimer` plugin (as not only minutes but also hours can be set).

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8240)

### CODE REVIEW

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/2cc072b40b99f95ebf891ba60d28f843a6d326ab/plugins/readtimer.koplugin/main.lua#L143
`(hours:minutes)` doesn't fit in the title. It can be moved to the menu item text
https://github.com/koreader/koreader/blob/2cc072b40b99f95ebf891ba60d28f843a6d326ab/plugins/readtimer.koplugin/main.lua#L125
because we do not have left/right texts like in the DoubleSpinWidget.

**Comment by hius07** [comment]
Is it worth to add the third numberpicker for seconds? (Not actually useful, but might look good)

**Comment by Frenzie** [comment]
What is the current behavior with seconds?

**Comment by hius07** [comment]
There are no seconds at all.

**Comment by Frenzie** [comment]
Yes, but what is done with the seconds when the time is set? Does it take the then current seconds or does it take the old seconds.

**Comment by hius07** [comment]
All devices support seconds, eg
https://github.com/koreader/koreader/blob/d3213e6b805b50cc07c021ea089681e36ebd4284/frontend/device/kindle/device.lua#L156
But we do not use seconds
https://github.com/koreader/koreader/blob/3a01ab78994af9d232b2cd584eec6d5ea8a05b76/frontend/ui/elements/common_settings_menu_table.lua#L140

**Comment by Frenzie** [comment]
Okay, and the behavior of date -s when seconds are absent is to just set it to `0`.

**Comment by zwim** [comment]
> `(hours:minutes)` doesn't fit in the title. It can be moved to the menu item text

I think it is clearer to add `info_text` to the `datetimewidget`. (And the menu entries are shorter :)

**Comment by zwim** [comment]
> Is it worth to add the third numberpicker for seconds? (Not actually useful, but might look good, and shorten the code)

I think a seconds granularity is too hectic. I/we want to read and not to rush ;-)

So I would not implement the seconds here. 

The code would only get 4 lines shorter, but a bit more complicated.

File: plugins/readtimer.koplugin/main.lua [hunk 1]
```diff
@@ -80,17 +80,19 @@ function ReadTimer:addToMainMenu(menu_items)
         end,
         sub_item_table = {
             {
-                text = _("Time"),
+                text = _("Point in Time"),
```
Comment (Frenzie) [inline plugins/readtimer.koplugin/main.lua:83]: Could you explain why you want to change this?

(As a generic remark for the future, that should be `Point in time`.)

File: plugins/readtimer.koplugin/main.lua [hunk 2]
```diff
@@ -80,17 +80,19 @@ function ReadTimer:addToMainMenu(menu_items)
...
                     local curr_hour = now_t.hour
                     local curr_min = now_t.min
-                    local time_widget = TimeWidget:new{
+                    local time_widget = DateTimeWidget:new{
+                        is_date = false,
                         hour = curr_hour,
                         min = curr_min,
                         ok_text = _("Set timer"),
                         title_text =  _("Set reader timer"),
+                        info_text = _("Enter time in hours an minutes."),
```
Comment (Frenzie) [inline plugins/readtimer.koplugin/main.lua:95]:
```suggestion
                        title_text =  _("New alarm"),
```

File: plugins/readtimer.koplugin/main.lua [hunk 3]
```diff
@@ -133,12 +135,14 @@ function ReadTimer:addToMainMenu(menu_items)
...
-                    local time_widget = TimeWidget:new{
+                    local time_widget = DateTimeWidget:new{
+                        is_date = false,
                         hour = remain_hours or 0,
                         min = remain_minutes or 0,
                         hour_max = 17,
                         ok_text = _("Set timer"),
-                        title_text =  _("Set reader timer from now (hours:minutes)"),
+                        title_text =  _("Set reader timer from now"),
+                        info_text = _("Enter time in hours an minutes."),
```
Comment (Frenzie) [inline plugins/readtimer.koplugin/main.lua:145]:
```suggestion
                        title_text =  _("Minutes from now"),
```

File: plugins/readtimer.koplugin/main.lua (hunk 1 above)
Comment (zwim) [reply plugins/readtimer.koplugin/main.lua:83]: The idea is that in the menu it is called read timer (in my understanding a timer is something similar to a countdown)
In the menu we really have such a timer (counting down) and additionaly an alarm clock.

By explicitly naming both entries as what they do, it is clear to the user what to select.

I think it is also easier for translators by naming the entries explicitly. If we call the first `alarm`, I would assume a German translator to use the German word `Alarm` which is not good at all.

File: plugins/readtimer.koplugin/main.lua (hunk 2 above)
Comment (zwim) [reply plugins/readtimer.koplugin/main.lua:95]: Is it really clear to any user that something like
   
grafik

shall be 00:00 (0 h 0 min)?

Or wouldn't a info text be more user friendly, something like

grafik

File: plugins/readtimer.koplugin/main.lua (hunk 1 above)
Comment (zwim) [reply plugins/readtimer.koplugin/main.lua:83]: BTW: Shouldn't the plugin's menu entry called like something: `Read timer and alarm`? A timer is similar to an alarm but not quite the same?

Comment (hius07) [reply plugins/readtimer.koplugin/main.lua:83]: Plugins menu entry: `Timer`
Submenu: `Set time`, `Set interval`, `Stop timer`.

Comment (Frenzie) [reply plugins/readtimer.koplugin/main.lua:83]: @zwim 
> I think it is also easier for translators by naming the entries explicitly. If we call the first alarm, I would assume a German translator to use the German word Alarm which is not good at all.

NB I implemented contexts for a reason.

@hius07 +1 on something like that

**Comment by poire-z** [comment]
Is this good to go ? And checked with all other bits of code that may use this widget, if any ?

File: plugins/readtimer.koplugin/main.lua [hunk 4]
```diff
@@ -133,12 +135,14 @@ function ReadTimer:addToMainMenu(menu_items)
...
-                    local time_widget = TimeWidget:new{
+                    local time_widget = DateTimeWidget:new{
+                        is_date = false,
                         hour = remain_hours or 0,
                         min = remain_minutes or 0,
                         hour_max = 17,
                         ok_text = _("Set timer"),
-                        title_text =  _("Set reader timer from now (hours:minutes)"),
+                        title_text =  _("Set reader timer"),
+                        info_text = _("Enter time in hours an minutes."),
```
Comment (Frenzie) [inline plugins/readtimer.koplugin/main.lua:145]: and

File: plugins/readtimer.koplugin/main.lua [hunk 5]
```diff
@@ -80,17 +80,19 @@ function ReadTimer:addToMainMenu(menu_items)
...
-                    local time_widget = TimeWidget:new{
+                    local time_widget = DateTimeWidget:new{
+                        is_date = false,
                         hour = curr_hour,
                         min = curr_min,
-                        ok_text = _("Set timer"),
-                        title_text =  _("Set reader timer"),
+                        ok_text = _("Set alarm"),
+                        title_text =  _("New alarm"),
+                        info_text = _("Enter time in hours an minutes."),
```
Comment (Frenzie) [inline plugins/readtimer.koplugin/main.lua:95]:
```suggestion
                        info_text = _("Enter a time in hours and minutes."),
```

**Comment by zwim** [comment]
> Is this good to go ? And checked with all other bits of code that may use this widget, if any ?

If have searche for `datewidget` and `timewidget` an only found them in `frontend/ui/elements/common_settings_menu_table.lua` and `plugins/readtimer.koplugin/main.lua`. There the new `datetimewidget` works fine.

**Pull request merged**